In [1]:
import numpy as np 
import pandas as pd

In [4]:
titanic_data = pd.read_csv("Titanic Dataset.csv")
titanic_data.head()

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500   NaN        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S  
3      0            113803  53.1000  C123        S  
4      0            373450   8.0500   NaN        S

In [5]:
titanic_data.shape

(891, 12)

# Machine Learning Algorithms

### Logistic Regression

In [7]:
# Data Preparation
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import classification_report
from sklearn.metrics import confusion_matrix,accuracy_score
from sklearn.impute import SimpleImputer
from sklearn.preprocessing import MinMaxScaler
cols = ['Pclass','Sex','Age','SibSp','Parch','Fare','Embarked']
scaler = MinMaxScaler()
X = titanic_data[cols]
y = titanic_data['Survived']
en_data = pd.get_dummies(X)

my_imputer = SimpleImputer()
imp_data = my_imputer.fit_transform(en_data)

x_train, x_test, y_train, y_test = train_test_split(imp_data, y, test_size=0.10, random_state=1)
logmodel = LogisticRegression()
logmodel.fit(x_train, y_train)

predictions = logmodel.predict(x_test)

print("accuracy_score = ",accuracy_score(y_test, predictions))
print("confusion_matrix = \n",confusion_matrix(y_test, predictions))
print("classification_report = \n",classification_report(y_test, predictions))

accuracy_score =  0.7777777777777778
confusion_matrix = 
 [[45  8]
 [12 25]]
classification_report = 
               precision    recall  f1-score   support

           0       0.79      0.85      0.82        53
           1       0.76      0.68      0.71        37

    accuracy                           0.78        90
   macro avg       0.77      0.76      0.77        90
weighted avg       0.78      0.78      0.78        90



C:\Users\Testbook\AppData\Local\Packages\PythonSoftwareFoundation.Python.3.8_qbz5n2kfra8p0\LocalCache\local-packages\Python38\site-packages\sklearn\linear_model\_logistic.py:444: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


### Decision Tree

In [8]:
from sklearn.tree import DecisionTreeClassifier
clf = DecisionTreeClassifier()
clf = clf.fit(x_train, y_train)
predictions = clf.predict(x_test)

print("accuracy_score = ",accuracy_score(y_test, predictions))
print("confusion_matrix = \n",confusion_matrix(y_test, predictions))
print("classification_report = \n",classification_report(y_test, predictions))

accuracy_score =  0.7333333333333333
confusion_matrix = 
 [[44  9]
 [15 22]]
classification_report = 
               precision    recall  f1-score   support

           0       0.75      0.83      0.79        53
           1       0.71      0.59      0.65        37

    accuracy                           0.73        90
   macro avg       0.73      0.71      0.72        90
weighted avg       0.73      0.73      0.73        90



### Random Forest Classifier

In [10]:
from sklearn.ensemble import RandomForestClassifier


rfc=RandomForestClassifier(n_estimators=100)
rfc.fit(x_train,y_train)

predictions = rfc.predict(x_test)

print("accuracy_score = ",accuracy_score(y_test, predictions))
print("confusion_matrix = \n",confusion_matrix(y_test, predictions))
print("classification_report = \n",classification_report(y_test, predictions))


accuracy_score =  0.7666666666666667
confusion_matrix = 
 [[45  8]
 [13 24]]
classification_report = 
               precision    recall  f1-score   support

           0       0.78      0.85      0.81        53
           1       0.75      0.65      0.70        37

    accuracy                           0.77        90
   macro avg       0.76      0.75      0.75        90
weighted avg       0.77      0.77      0.76        90



From Above we can conclude that Logistic Regression is the Best Model for the Data , Having Accuracy of  0.7777777777777778